src/main/java/org/jabref/logic/openoffice/DocumentConnection.java

```diff
@@ -0,0 +1,1305 @@
+
+        XDocumentPropertiesSupplier supp = unoQI(XDocumentPropertiesSupplier.class, mxDoc);
+        this.userProperties = supp.getDocumentProperties().getUserDefinedProperties();
+        this.propertySet = unoQI(XPropertySet.class, userProperties);
+    }
+
+    /**
+     * unoQI : short for UnoRuntime.queryInterface
+     *
+     * @return A reference to the requested UNO interface type if available,
+     *         otherwise null
+     */
+    private static <T> T unoQI(Class<T> zInterface,
+                               Object object) {
+        return UnoRuntime.queryInterface(zInterface, object);
+    }
+
+    private XDocumentProperties getXDocumentProperties() {
+        XDocumentPropertiesSupplier supp = unoQI(XDocumentPropertiesSupplier.class, mxDoc);
+        return supp.getDocumentProperties();
+    }
+
+    private short getRedlineDisplayType()
+        throws
+        UnknownPropertyException,
+        WrappedTargetException {
+        // https://www.openoffice.org/api/docs/common/ref/com/sun/star/document/RedlineDisplayType.html
+        // Constants (short)
+        // 0 NONE                  no changes are displayed.
+        // 1 INSERTED              only inserted parts are displayed and attributed.
+        // 2 INSERTED_AND_REMOVED  only inserted parts are displayed and attributed.
+        // 3 REMOVED               only removed parts are displayed and attributed.
+        XPropertySet ps = unoQI(XPropertySet.class, mxDoc);
+        return (short) ps.getPropertyValue("RedlineDisplayType");
+    }
+
+    public boolean getRecordChanges()
+        throws
+        UnknownPropertyException,
+        WrappedTargetException {
+        XPropertySet ps = unoQI(XPropertySet.class, mxDoc);
+        // https://wiki.openoffice.org/wiki/Documentation/DevGuide/Text/Settings
+        // "Properties of com.sun.star.text.TextDocument"
+        if (ps == null) {
+            throw new RuntimeException("getRecordChanges: ps is null");
+        }
+        return (boolean) ps.getPropertyValue("RecordChanges");
+    }
+
+    public void setRecordChanges(boolean value)
+        throws
+        UnknownPropertyException,
+        WrappedTargetException,
+        PropertyVetoException {
+        XPropertySet ps = unoQI(XPropertySet.class, mxDoc);
+        ps.setPropertyValue("RecordChanges", value);
+    }
+
+    private XRedlinesSupplier getRedlinesSupplier() {
+        return unoQI(XRedlinesSupplier.class, mxDoc);
+    }
+
+    public int countRedlines() {
+        XRedlinesSupplier rs = getRedlinesSupplier();
+        XEnumerationAccess ea = rs.getRedlines();
+        XEnumeration e = ea.createEnumeration(); // null for empty
+        if (e == null) {
+            // System.out.println("countRedLines: no redlines found");
+            return 0;
+        } else {
+            int count = 0;
+            for (; e.hasMoreElements(); ) {
+                try {
+                    Object o = e.nextElement();
```

Comment:
Very minor. Prefer `while` loops with this kind of statement. You also shouldn't need to assign `Object o = e.nextElement();`, just leaving it as `e.nextElement()` is fine.

Reply:
ok